src/hackney_http.erl

```diff
@@ -325,7 +328,40 @@ parse_trailers(St, Acc) ->
     _ -> error
   end.
 
-parse_body(St=#hparser{body_state=waiting, te=TE, clen=Length,
+-define(MAX_WBITS, 15).	% zconf.h
+parse_body(St=#hparser{body_state=waiting, ce=CE, clen=Length, method=Method}) ->
+  St2 = case CE of
+    _ when Length =:= 0 orelse Method =:= <<"HEAD">> ->
+      St;
+    <<"gzip">> ->
+      Z = zlib:open(),
+      ok = zlib:inflateInit(Z, 16 + ?MAX_WBITS),	% http://www.zlib.net/manual.html#Advanced
+      St#hparser{encoding={zlib,Z}};
+    <<"deflate">> ->
+      Z = zlib:open(),
+      ok = zlib:inflateInit(Z, ?MAX_WBITS),
+      St#hparser{encoding={zlib,Z}};
+    _ ->
+      St
+  end,
+  parse_body2(St2);
+parse_body(St=#hparser{encoding={zlib,Z}}) ->
+  case parse_body2(St) of
+    {ok, Chunk, St2} ->
+      Chunk2 = iolist_to_binary(zlib:inflate(Z, Chunk)),
+      {ok, Chunk2, St2};
+    {done, Rest} ->
+      Rest2 = iolist_to_binary(zlib:inflate(Z, Rest)),
+      ok = zlib:inflateEnd(Z),
+      ok = zlib:close(Z),
```

Comment:
Are you sure this is the only condition when zlib handle might be closed? Is it possible to leave some zlib handles stale in case of errors or smth like that?

Reply:
Nope, pretty sure we leak if something implodes (ie. server prematurely closes stream).

Really at this point more interested in:
 * does this work for others
 * anyone see any regressions
 * does `compress` as an option suit the users of hackney
 * is there a better layer for me to plumb this in or less intrusively

Fixing bugs/edge cases I suspect would not change drastically the shape of this patch but changing the plumbing would, as would maybe some fixes for regressions.  Make sense?

I dislike wasting time if I am on the wrong footpath and no one tells me :)

Reply:
"does compress as an option suit the users of hackney. " yes